File: tests/python/test_e2e_parity.py
"""End-to-end parity test exercising the full SDK shape.

This scenario is mirrored in the Node and WASM test suites to verify
cross-binding behavioral parity.

Covers:
- A 2-stage Pipeline built via PipelineBuilder.
- A custom CompletionModel subclass (sync override of complete) that
  returns a deterministic response without touching the network.
- A PromptTemplate rendered with the previous stage's output.
- A Memory store backed by InMemoryBackend, written to and queried.

No network access or API keys are required.
"""

import pytest

from blazen import (
    ChatMessage,
    CompletionModel,
    Context,
    Event,
    InMemoryBackend,
    Memory,
    Pipeline,
    PipelineBuilder,
    PipelineResult,
    PromptTemplate,
    Stage,
    StageResult,
    StartEvent,
    StopEvent,
    TemplateRole,
    Workflow,
    step,
)


# Deterministic mock provider used in stage 1.
class HelloWorldLLM(CompletionModel):
    """A CustomProvider subclass that always returns 'Hello World'."""

    def __new__(cls):
        return super().__new__(cls, model_id="hello-world-mock")

    async def complete(self, messages, options=None):
        # Sync logic wrapped in an async def -- no awaiting, no IO.
        return {
            "content": "Hello World",
            "model": "hello-world-mock",
            "tool_calls": [],
            "finish_reason": "stop",
        }


@pytest.mark.asyncio
async def test_e2e_parity_pipeline_with_memory_and_prompt_template():
    # Single shared instances used across stages.
    llm = HelloWorldLLM()
    backend = InMemoryBackend()
    memory = Memory.local(backend)
    template = PromptTemplate(
        "Echo says: {{phrase}}",
        role=TemplateRole.User,
        name="echo-template",
    )

    # Stage 1: ask the custom provider for a deterministic completion.
    @step
    async def call_llm(ctx: Context, ev: StartEvent) -> StopEvent:
        response = await llm.complete([ChatMessage.user("greet me")])
        # Subclass returns a dict, accessible by key.
        content = response["content"]
        return StopEvent(result={"llm_text": content})

    # Stage 2: render a PromptTemplate from stage 1 output and persist
    # the rendered text to Memory.
    @step
    async def render_and_store(ctx: Context, ev: StartEvent) -> StopEvent:
        data = ev.to_dict()
        phrase = data.get("llm_text", "")
        rendered = template.render(phrase=phrase)
        # Persist the rendered prompt content into memory.
        doc_id = await memory.add("rendered-doc", rendered.content)
        return StopEvent(
            result={
                "rendered": rendered.content,
                "doc_id": doc_id,
                "role": rendered.role,
            }
        )

    wf_llm = Workflow("llm-wf", [call_llm])
    wf_render = Workflow("render-wf", [render_and_store])

    stage_llm = Stage(name="generate", workflow=wf_llm)
    stage_render = Stage(name="render", workflow=wf_render)

    pipeline = (
        PipelineBuilder("e2e-parity")
        .stage(stage_llm)
        .stage(stage_render)
        .build()
    )
    assert isinstance(pipeline, Pipeline)

    # Run the pipeline.
    handler = await pipeline.start()
    result = await handler.result()

    # PipelineResult shape assertions.
    assert isinstance(result, PipelineResult)
    assert result.pipeline_name == "e2e-parity"

    stage_results = result.stage_results
    assert len(stage_results) == 2
    for sr in stage_results:
        assert isinstance(sr, StageResult)
        assert sr.skipped is False

    # Stage 1: custom provider produced "Hello World".
    assert stage_results[0].name == "generate"
    assert stage_results[0].output == {"llm_text": "Hello World"}

    # Stage 2: PromptTemplate rendered correctly with stage 1 result.
    assert stage_results[1].name == "render"
    final = stage_results[1].output
    assert final["rendered"] == "Echo says: Hello World"
    assert final["doc_id"] == "rendered-doc"
    assert final["role"] == "user"

    # final_output of the pipeline is the last stage's output.
    assert result.final_output["rendered"] == "Echo says: Hello World"

    # Memory backend was written to: count, get, search_local all succeed.
    count = await memory.count()
    assert count == 1

    entry = await memory.get("rendered-doc")
    assert entry is not None
    assert entry["text"] == "Echo says: Hello World"

    hits = await memory.search_local("Hello World", limit=5)
    assert isinstance(hits, list)
    assert len(hits) >= 1
    assert any(h.id == "rendered-doc" for h in hits)
    assert any(h.text == "Echo says: Hello World" for h in hits)


# Reference imports kept to assert the symbols are exported.
_ = (Event,)

File: tests/node/test_e2e_parity.mjs
// End-to-end parity test for the Node.js bindings.
//
// Mirrors tests/python/test_e2e_parity.py: verifies that a 2-stage Pipeline
// composed of Stage + Workflow + CustomProvider + InMemoryBackend-backed
// Memory + PromptTemplate can be wired together via the public napi-rs
// surface without throwing at construction time, and that the resulting
// Pipeline exposes the expected method shape.
//
// Test bodies are SYNC ONLY -- per project memory async test bodies hang
// the Node test runner because the napi-rs tokio runtime keeps handles
// alive past Promise resolution. Pipeline.start() is therefore never
// invoked from this file; runtime behavior is exercised separately by
// subprocess-based harnesses.
//
// Build the native binding first:
//   pnpm --filter blazen run build

import test from "ava";

import {
  ChatMessage,
  CompletionModel,
  CustomProvider,
  EmbeddingModel,
  InMemoryBackend,
  Memory,
  Pipeline,
  PipelineBuilder,
  PromptTemplate,
  Stage,
  Workflow,
} from "../../crates/blazen-node/index.js";

// ===========================================================================
// Helpers -- minimal stub subclasses used purely for shape wiring.
// ===========================================================================

// A trivial CompletionModel subclass. complete() is never invoked from this
// file so the default "subclass must override" body is fine.
class StubLLM extends CompletionModel {
  constructor() {
    super({ modelId: "stub-llm", contextLength: 4096 });
  }
}

// A trivial EmbeddingModel subclass. We do NOT pass this into Memory because
// Memory rejects subclassed embedders without an inner provider (see
// test_subclass_providers.mjs). For the Memory branch we use Memory.local()
// which requires no embedder at all.
class StubEmbed extends EmbeddingModel {
  constructor() {
    super({ modelId: "stub-embed", dimensions: 16 });
  }
}

// ===========================================================================
// e2e parity (shape)
// ===========================================================================

test("e2e parity (shape) · Pipeline + CustomProvider + InMemoryBackend + PromptTemplate wire together", (t) => {
  // 1. PromptTemplate -- exercise the parity primitive.
  const template = new PromptTemplate(
    "Summarize the following text: {{text}}",
    { role: "user", name: "summarize" },
  );
  t.is(template.role, "user");
  t.is(template.name, "summarize");
  t.deepEqual(template.variables, ["text"]);

  const rendered = template.render({ text: "hello world" });
  t.truthy(rendered instanceof ChatMessage);
  t.is(rendered.role, "user");
  t.truthy(rendered.content && rendered.content.includes("hello world"));

  // 2. CustomProvider -- wrap a host object exposing a TTS-shaped method.
  // No method is invoked here; we only verify the wrapper constructs.
  const host = {
    async textToSpeech(_request) {
      return { audio: [], timing: {}, metadata: {} };
    },
  };
  const customProvider = new CustomProvider(host, {
    providerId: "stub-custom",
  });
  t.is(customProvider.providerId, "stub-custom");

  // 3. InMemoryBackend wrapped in a Memory (local-only mode -- no
  // embedder needed, and avoids the "subclassed EmbeddingModel" rejection
  // that Memory's full constructor enforces).
  const backend = new InMemoryBackend();
  const memory = Memory.local(backend);
  t.truthy(memory instanceof Memory);

  // Sanity: ensure StubEmbed/StubLLM also construct so their inclusion
  // in this parity test mirrors the Python file's surface, even though
  // we don't hand them to Memory.
  const embedder = new StubEmbed();
  const llm = new StubLLM();
  t.is(embedder.modelId, "stub-embed");
  t.is(llm.modelId, "stub-llm");

  // 4. Build two distinct workflows and wrap them as Stages.
  const wf1 = new Workflow("stage-1");
  wf1.addStep("ingest", ["blazen::StartEvent"], (_event, _ctx) => {
    // Hand off to stage 2 via StopEvent containing prompt text.
    return {
      type: "blazen::StopEvent",
      result: { prompt: "hello world" },
    };
  });

  const wf2 = new Workflow("stage-2");
  wf2.addStep("respond", ["blazen::StartEvent"], (_event, _ctx) => {
    return {
      type: "blazen::StopEvent",
      result: { reply: "ok" },
    };
  });

  const stage1 = new Stage("ingest", wf1);
  const stage2 = new Stage("respond", wf2);
  t.is(stage1.name, "ingest");
  t.is(stage2.name, "respond");

  // 5. Build the 2-stage Pipeline.
  const pipeline = new PipelineBuilder("e2e-parity")
    .stage(stage1)
    .stage(stage2)
    .build();

  t.truthy(pipeline instanceof Pipeline);
  t.is(typeof pipeline.start, "function");
  t.is(typeof pipeline.resume, "function");
});

test("e2e parity (shape) · PromptTemplate role override flows through to ChatMessage", (t) => {
  const sysTemplate = new PromptTemplate(
    "You are a helpful assistant named {{name}}.",
    { role: "system", name: "system-prompt" },
  );
  t.is(sysTemplate.role, "system");

  const msg = sysTemplate.render({ name: "Blazen" });
  t.is(msg.role, "system");
  t.truthy(msg.content && msg.content.includes("Blazen"));
});

test("e2e parity (shape) · PipelineBuilder rejects empty pipelines (parity with Python)", (t) => {
  // Mirrors the Python parity assertion that an empty pipeline cannot be
  // built -- the builder must require at least one stage.
  t.throws(
    () => new PipelineBuilder("empty").build(),
    { message: /stage|empty|at least/i },
  );
});

test("e2e parity (shape) · CustomProvider providerId defaults to 'custom' when omitted", (t) => {
  // Mirrors the Python parity expectation that an unspecified provider id
  // falls back to a stable default.
  const provider = new CustomProvider({});
  t.is(provider.providerId, "custom");
});
